### ISSUE TITLE
[Feature]: Fetch groups information from OIDC UserInfo endpoint

### ISSUE THREAD

**Issue by ckolbitsch-work** [oauth2-proxy/oauth2-proxy#2864]
### Motivation

Users may belong to many groups - more than may be able to be reasonably embedded in a JWT token in the "groups" claim.

We want to be able to configure oauth2-proxy in a way that it validates JWT token that the user provides the system (i.e., without oauth2-proxy triggering the flow, but the client already having performed that flow) and validates if user belongs to a required group.

In some OIDC providers, the user JWT token may contain a subset (or no) groups, if the user belongs to so many groups that it becomes unmanageable to embed all groups in the token.

We would like to be able to configure oauth2-proxy to detect that the user does not have the required group listed in the groups claim in the JWT token and to reach out to the OIDC _UserInfo_ endpoint to retrieve the full list of groups.

NOTE: I'm writing this specific to the OIDC provider, but most likely this applies to other providers as well.

### Possible solution

The OIDC client is already able to make requests to the OIDC provider, so it would "only" be issuing the additional request to the OIDC provider for retrieving group information as part of the JWT validation. Given that (at least in our scenario) the incoming request already has a JWT token, oauth2-proxy could even leverage that JWT for talking to the OIDC provider for requesting the _UserInfo_ (in case the client-id/-secret might not entitle them to do so).

For configuration, there could be a set of options to choose from: for example
- query the group membership only if there is no group claim (e.g., the OIDC provider might refuse to embed groups if there are too many)
- query the group membership only if the group is not listed in the groups claim (e.g., the OIDC provider chooses to embed a subset of groups if there are too many -- in this case, the "normal case" in which the user is part of the group and the list of groups is small, it would avoid the round-trip to the OIDC provider endpoint)
- query the group membership depending on some criteria in the JWT token (e.g., we work with an OIDC provider that has a claim called "number_of_groups". We would like to be able to check if the groups claim might contain a subset of groups and only request more groups if we know that the group was cut short)

Another implementation topic would be if this can be cached. Doing a round-trip for every request seem problematic. Given that the JWT token was supposed to contain all groups, it seems reasonable that oauth2-proxy may cache UserInfo data for a particular token for the duration of the token's validity (i.e., if the user's groups are extended, the token would need to be renewed for oauth2-proxy to perform another request - that would be similar to the case where all groups were included in the JWT token to begin with)

### Provider

oidc

**Comment by JoelSpeed** [comment]
I believe we already do something like this, with the `EnrichSession` function that each provider implements. I believe we call out there in a similar way to fetch missing information, so I guess if that isn't calling to add group information, then it would make sense to extend it to add additional group information as you've described

**Comment by kolbitsch-lastline** [comment]
yes, that looks like the right place to add this - unfortunately this is currently more or less an empty stub **for OIDC**:

https://github.com/oauth2-proxy/oauth2-proxy/blob/master/providers/oidc.go#L100

```
func (p *OIDCProvider) EnrichSession(_ context.Context, s *sessions.SessionState) error {
	// If a mandatory email wasn't set, error at this point.
	if s.Email == "" {
		return errors.New("neither the id_token nor the profileURL set an email")
	}
	return nil
}
```
